File: Directory/20. Valid Parentheses/Solution.py
# -*- coding: utf-8 -*-
# @Author: yufeng
# @Date:   2018-10-13 22:17:16
# @Last Modified by:   yufeng
# @Last Modified time: 2018-10-13 22:34:11
# @Email: fzhiy270@sina.com
# @Blog: http://fzhiy.com/

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pattern = {'(':')', '[':']', '{':'}'}
        stack = []
        for ch in s:
        	if ch in pattern.keys():
        		stack.append(ch)
        	elif not stack or ch!=pattern[stack.pop()]:# 栈空或者匹配非法
        		return False
        return not stack

File: Directory/20. Valid Parentheses/Solution.java
import java.util.Stack;

public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for (char ch : s.toCharArray()) {
            if (ch == '(') {
                stack.push(')');
            } else if (ch == '[') {
                stack.push(']');
            } else if (ch == '{') {
                stack.push('}');
            } else if (stack.empty() || stack.pop() != ch) {///栈为空或者栈顶不等于ch
                return false;
            }
        }
        return stack.empty();
    }

    public static void main(String[] args) {
        String str = "" + "()";
        boolean res = new Solution().isValid(str);
        System.out.println(res);
    }
}


